agent/agent.go

```diff
@@ -205,6 +209,153 @@ func (a *agent) runLoop(ctx context.Context) {
 	}
 }
 
+func (a *agent) collectMetadata(ctx context.Context, md codersdk.WorkspaceAgentMetadataDescription) *codersdk.WorkspaceAgentMetadataResult {
+	timeout := md.Timeout
+	if timeout == 0 {
+		timeout = md.Interval
+	}
+	ctx, cancel := context.WithDeadline(ctx, time.Now().Add(
+		time.Duration(timeout)*time.Second,
+	),
+	)
+	defer cancel()
+
+	var out bytes.Buffer
+
+	result := &codersdk.WorkspaceAgentMetadataResult{
+		// CollectedAt is set here for testing purposes and overrided by
+		// the server to the time the server received the result to protect
+		// against clock skew.
+		//
+		// In the future, the server may accept the timestamp from the agent
+		// if it is certain the clocks are in sync.
+		CollectedAt: time.Now(),
+	}
+	cmd, err := a.createCommand(ctx, md.Script, nil)
+	if err != nil {
+		result.Error = err.Error()
+		return result
+	}
+
+	cmd.Stdout = &out
+	cmd.Stderr = &out
+
+	// The error isn't mutually exclusive with useful output.
+	err = cmd.Run()
+
+	const bufLimit = 10 << 10
+	if out.Len() > bufLimit {
+		err = errors.Join(
+			err,
+			xerrors.Errorf("output truncated from %v to %v bytes", out.Len(), bufLimit),
+		)
+		out.Truncate(bufLimit)
+	}
+
+	if err != nil {
+		result.Error = err.Error()
+	}
+	result.Value = out.String()
+	return result
+}
+
+func convertInterval(i int64) time.Duration {
+	// In tests we want to set shorter intervals because engineers are
+	// impatient.
+	base := time.Second
+	if flag.Lookup("test.v") != nil {
+		base = time.Millisecond * 100
+	}
+	return time.Duration(i) * base
```

Comment:
It seems likely to me that some may want to set sub-second intervals as well, in which case this abstraction is bad anyways...

Reply:
See https://github.com/coder/terraform-provider-coder/issues/117

While I think we should accept Duration strings for clarity, I don't think we should accept sub-second intervals because of the DB write load. Subsecond timeouts are reasonable though. If a `ps aux` is taking more than a 100ms you might not want it to happen....